enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,108 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+creation-date: 2022-01-31
+last-updated: 2022-01-31
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM malfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect inactive cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
```

Comment:
> we need a controller process that will run close to the Kubernetes API server
Can you help me understand the goal better - it feels like this controller can be a test that runs in CI only (it does not need to run in production cluster)
The feature it is testing is - operators should write to `status.[]conditions`

The controller can be a test that goes through all `status.[]conditions` with a very aggressive interval, let's say `3m` and sets the message to `[stale marker] + {original message}` and if the message does not get reverted to its original form we can fail the test on this operator

This test establishes the invariant: each `co` backed operator writes to all `status.[]conditions`.

Now, if we see stale conditions in a real life cluster, we don't have to guess whether the operator is buggy in writing to `status.[]conditions`. the staleness is an indication of an underlying condition, like api connectivity.

Reply:
> Now, if we see stale conditions in a real life cluster, we don't have to guess whether the operator is buggy in writing to status.[]conditions. the staleness is an indication of an underlying condition, like api connectivity.

the problem is how do we recognize stale conditions in a real life cluster?  If the condition is set to "available=true" and never changes, is that accurate, or is that a stale condition due to a buggy/broken/failed operator?

CI detection isn't enough because we're not trying to prove that "in the happy path/normal case, an operator updates its conditions".  We're trying to detect scenarios where the operator has stopped doing so for unanticipated reasons.

(put another way, the operator could still be buggy, in a way that CI does not expose)